File: backup/golang/medium/46.全排列.go
package medium

/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 *
 * https://leetcode-cn.com/problems/permutations/description/
 *
 * algorithms
 * Medium (70.87%)
 * Likes:    379
 * Dislikes: 0
 * Total Accepted:    46.5K
 * Total Submissions: 64.8K
 * Testcase Example:  '[1,2,3]'
 *
 * 给定一个没有重复数字的序列，返回其所有可能的全排列。
 *
 * 示例:
 *
 * 输入: [1,2,3]
 * 输出:
 * [
 * ⁠ [1,2,3],
 * ⁠ [1,3,2],
 * ⁠ [2,1,3],
 * ⁠ [2,3,1],
 * ⁠ [3,1,2],
 * ⁠ [3,2,1]
 * ]
 *
 */
func permute(nums []int) [][]int {
	// 迭代的方式
	res := [][]int{}
	if len(nums) == 0 {
		return res
	} else if len(nums) == 1 {
		res = append(res, nums)
		return res
	}

	for idx := 0; idx < len(nums); idx++ {
		last := []int{}
		last = append(last, nums[:idx]...)
		last = append(last, nums[idx+1:]...)
		for _, ans := range permute(last) {
			t := []int{nums[idx]}
			t = append(t, ans...)
			res = append(res, t)
		}
	}
	return res

}

File: labuladong/开始之前/backtracking/46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
from copy import copy


class Solution:
    def permute(self, nums):
        res = []

        def backtrack(now, last):
            if len(last) == 0:
                res.append(now)
                return

            for num in last:
                old = copy(now)  # 备份一下当前的选择
                now.append(num)  # 做选择
                last = last[1:]
                backtrack(now, last)
                now = old  # 回退选择
                last.append(num)  # 回退选择

        backtrack([], nums)
        return res

    def test(self, nums=[1, 2, 3]):
        return self.permute(nums)


# @lc code=end
